turbo/jsonrpc/eth_call.go

```diff
@@ -334,100 +337,160 @@ func (api *APIImpl) EstimateGas(ctx context.Context, argsOrNil *ethapi2.CallArgs
+	if err != nil {
+		return nil, err
+	} else if blockNr == 0 {
+		return nil, errors.New("block not found")
+	}
 
-			hashStageCfg := stagedsync.StageHashStateCfg(nil, api.dirs, api.historyV3(batch))
-			if err := stagedsync.UnwindHashStateStage(unwindState, stageState, batch, hashStageCfg, ctx, api.logger); err != nil {
-				return nil, err
-			}
+	// get the root hash from header to validate proofs along the way
+	header, err := api._blockReader.HeaderByNumber(ctx, roTx, blockNr)
+	if err != nil {
+		return nil, err
+	}
 
-			interHashStageCfg := stagedsync.StageTrieCfg(nil, false, false, false, api.dirs.Tmp, api._blockReader, nil, api.historyV3(batch), api._agg)
-			loader, err = stagedsync.UnwindIntermediateHashesForTrieLoader("eth_getProof", rl, unwindState, stageState, batch, interHashStageCfg, nil, nil, ctx.Done(), api.logger)
-			if err != nil {
-				return nil, err
+	roTx2, err := db.BeginRo(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer roTx2.Rollback()
+	txBatch2 := membatchwithdb.NewMemoryBatch(roTx2, "", logger)
+	defer txBatch2.Rollback()
+
+	domains, err := libstate.NewSharedDomains(txBatch2, log.New())
+	if err != nil {
+		return nil, err
+	}
+	sdCtx := libstate.NewSharedDomainsCommitmentContext(domains, commitment.ModeUpdate, commitment.VariantHexPatriciaTrie)
+	patricieTrie := sdCtx.Trie()
+	hph, ok := patricieTrie.(*commitment.HexPatriciaHashed)
+	if !ok {
+		return nil, errors.New("casting to HexPatriciaTrieHashed failed")
+	}
+
+	// define these keys as "updates", but we are not really updating anything, we just want to load them into the grid,
+	// so this is just to satisfy the current hex patricia trie api.
+	updates := commitment.NewUpdates(commitment.ModeDirect, sdCtx.TempDir(), hph.HashAndNibblizeKey)
+	// touch account
+	updates.TouchPlainKey(string(address.Bytes()), nil, updates.TouchAccount)
+	// touch storage keys
+	for _, storageKey := range storageKeys {
+		updates.TouchPlainKey(string(common.FromHex(address.Hex()[2:]+storageKey.String()[2:])), nil, updates.TouchStorage)
+	}
+	hph.SetTrace(false)
+
+	// generate the trie for proofs, this works by loading the merkle paths to the touched keys
+	proofTrie, proofRootHash, err := hph.GenerateProofTrie(ctx, updates, header.Root[:], "eth_getProof")
+	if err != nil {
+		return nil, err
+	}
+
+	// verify hash
+	if !bytes.Equal(proofRootHash, header.Root[:]) {
+		return nil, fmt.Errorf("proof root hash mismatch actual(%x)!=expected(%x)", proofRootHash, header.Root[:])
+	}
+
+	a := new(big.Int).SetInt64(0)
+	// set initial response fields
+	proof := &accounts.AccProofResult{
+		Address:      address,
+		Balance:      (*hexutil.Big)(a),
+		Nonce:        hexutil.Uint64(0),
+		CodeHash:     libcommon.Hash{},
+		StorageHash:  libcommon.Hash{},
+		StorageProof: make([]accounts.StorProofResult, len(storageKeys)),
+	}
+
+	// get account proof
+	accountProof, _, err := proofTrie.Prove(crypto.Keccak256(address.Bytes()), 0, false)
```

Comment:
what actually stops us from generating proof during fold?

Reply:
i mean besides of ready-to-go api? do we lack some information during fold phase for that?